File: backend/python/app/schemas/timeline.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.timeline_event_type import TimelineEventSeverity, TimelineEventType, TimelinePeriod


class TimelineListQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=10, ge=1, le=100)
    search: str | None = None
    type: TimelineEventType | None = None
    service_id: int | None = Field(default=None, gt=0)
    incident_id: int | None = Field(default=None, gt=0)
    alert_id: int | None = Field(default=None, gt=0)
    period: TimelinePeriod = TimelinePeriod.all


class TimelineEventResponse(BaseModel):
    id: str
    timestamp: datetime
    type: TimelineEventType
    title: str
    description: str
    service_id: int | None = None
    incident_id: int | None = None
    alert_id: int | None = None
    severity: TimelineEventSeverity | None = None


class TimelineStatsResponse(BaseModel):
    events_today: int
    incidents: int
    alerts: int
    service_events: int


class PaginatedTimelineResponse(BaseModel):
    items: list[TimelineEventResponse]
    page: int
    page_size: int
    total: int
    total_pages: int
    stats: TimelineStatsResponse

File: frontend/src/types/api/timeline.ts
/** Backend timeline event types. */
export type ApiTimelineEventType =
  | 'incident_created'
  | 'incident_updated'
  | 'incident_resolved'
  | 'alert_triggered'
  | 'alert_acknowledged'
  | 'service_degraded'
  | 'service_recovered'
  | 'deployment'
  | 'maintenance'

/** Backend timeline event severity values. */
export type ApiTimelineEventSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info'

/** Backend timeline period filter values. */
export type ApiTimelinePeriod = 'all' | 'today' | '7d' | '30d'

export interface ApiTimelineEvent {
  id: string
  timestamp: string
  type: ApiTimelineEventType
  title: string
  description: string
  service_id: number | null
  incident_id: number | null
  alert_id: number | null
  severity: ApiTimelineEventSeverity | null
}

export interface ApiTimelineStats {
  events_today: number
  incidents: number
  alerts: number
  service_events: number
}

export interface PaginatedTimelineEvents {
  items: ApiTimelineEvent[]
  page: number
  page_size: number
  total: number
  total_pages: number
  stats: ApiTimelineStats
}

export interface TimelineListParams {
  page?: number
  page_size?: number
  search?: string
  type?: ApiTimelineEventType
  service_id?: number
  incident_id?: number
  alert_id?: number
  period?: ApiTimelinePeriod
}
